### PR TITLE
Osra315 hq cucumber tests for authentication

### PR DESCRIPTION
**Pull request by razorcd**
https://osraav.atlassian.net/browse/OSRA-315

**Notice:**
- Model specs are run together with the aa spec tests!
- Cucumber tests are running in verbose mode: it lists all the loaded files when starts
- features/support is shared between the cucumber:aa and cucumber:hq tests
- features/support/step_helper.rb has the helper methods for hq step definitions 
(features/support folder is always loaded first by cucumber)

**Commands:**
- `$rake`    # -> will run all tests: spec(aa and hq), `cucumber:hq` and `cucumber:aa`
- `$rake test:aa` # -> will run all aa tests: `spec:aa`, `cucumber:aa`
- `$rake test:hq` # -> will run all hq tests: `spec:hq`, `cucumber:hq`
- `$rake spec:aa` # -> will run all aa specs: all specs that are not in a 'hq' folder
- `$rake spec:hq` # -> will run all hq specs: all specs that are in a 'hq' folder
- `$rake cucumber:aa` # -> will run all aa cucumber tests: all features from features/aa and will load all .rb files from  features/aa and features/support
- `$rake cucumber:hq` # -> will run all hq cucumber tests: all features from features/hq and will load all .rb files from  features/hq and features/support

- `$cucumber` and `$cucumber -p aa` # -> will run features in 'aa' profile
- `$cucumber -p hq` # -> will run features in 'hq' profile
- `$rspec spec/controllers/hq/users_controller_spec.rb` $ -> will run rspec tests from one file
- `$cucumber -p aa features/aa/aa_features/users.feature` # -> will run one feature file in 'aa' profile
- `$cucumber -p hq features/hq/hq_features/authentication.feature` # -> will run one feature file in 'hq' profile

### CODE REVIEW

File: features_hq/support/step_helper.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+  def login email= DEFAULT_ADMIN[:email], password= DEFAULT_ADMIN[:password]
+    visit new_admin_user_session_path
+    fill_in "admin_user_email", :with => email
+    fill_in "admin_user_password", :with => password
+    click_button "Login"
+  end
+
+  def logout
+    visit destroy_admin_user_session_path
```
Comment (razorcd) [inline features_hq/support/step_helper.rb]: or should I use `session.clear` here?

File: features_hq/README.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
+Cucumber Features
```
Comment (PurityControl) [inline features_hq/README.md]: I wouldn't copy the readme. I would have 1 line referring to the authoritative source

File: features_hq/hq/authentication.feature [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+  Scenario: Restricted pages should redirect unauthenticated users to login page
+    Given I am an unauthenticated user
+    When I go to the "partners" page for the "admin" role
+    Then I should be on the "login" page for the "admin" role
+
+  Scenario: Restricted pages should be accessibe by authenticated users
+    Given I am a new, authenticated user
+    When I go to the "partners" page for the "admin" role
+    Then I should be on the "partners" page for the "admin" role
```
Comment (razorcd) [inline features_hq/hq/authentication.feature]: Is "admin" role used correctly? I am a little confused by this. We'll have "admin" and "user" roles in the future?

File: features_hq/support/env.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
+# IMPORTANT: This file is generated by cucumber-rails - edit at your own peril.
```
Comment (PurityControl) [inline features_hq/support/env.rb]: if we need to keep this in sync with active admin should they be symlinks?

File: Rakefile [hunk 5]
```diff
@@ -7,4 +7,5 @@ Osra::Application.load_tasks
 
 require 'coveralls/rake/task'
 Coveralls::RakeTask.new
-task :test_with_coveralls => [:spec, :cucumber, 'coveralls:push']
+#task :test_with_coveralls => [:cucumber_hq, 'coveralls:push']
+task :test_with_coveralls => [:spec, :cucumber_hq, :cucumber, 'coveralls:push']
```
Comment (razorcd) [inline Rakefile]: I get `[Coveralls] Outside the Travis environment, not sending data.` after running `$ rake test_with_coveralls`. 
This happens after adding  the `:cucumber_hq`  tests to this `:test_with_coveralls` task. 
Not sure what else I have to configure for coveralls.

**Comment by razorcd** [comment]
Should I create independent 
- './features/admin'  and  './features/hq'       or
- './features_admin'  and  './features_hq'

All features, steps and helper files independent.

File: features_hq/support/step_helper.rb (hunk 1 above)
Comment (theB264) [inline features_hq/support/step_helper.rb]: I would think in an integration test `session.clear` would not be the way to go, because an end-user cannot accomplish that directly

**Comment by NikitaAvvakumov** [comment]
I'd prefer separate subfolders within './features/'.

**Comment by razorcd** [comment]
@NikitaAvvakumov 
Looks like there are 2 semaphores running on our repo !?:
https://semaphoreapp.com/nikita_avvakumov/osra--4/branches/pull-request-261/builds/3
https://semaphoreapp.com/mots/osra/branches/pull-request-261/builds/3

**Comment by theB264** [comment]
Re: entire file at `features/hq/hq_features/authentication.feature`
I like the rest of this, especially using `features/hq` instead of `features_hq`.  But, could we rename the `hq_features` directory to something else (`access`?) or move the `authentication.feature` file up one directory if it will be the only authentication-related feature file.

**Comment by razorcd** [comment]
The `features/hq/hq_features` will hold all the future features too, not only the `authentication.feature`.
Not sure if by renaming it `access` will define it better. Or am I missing something?

Before this pr, the current `aa_features` folder was named `admin`.

**Comment by PurityControl** [comment]
have you tried to call an aa step from an hq feature file just to ensure that this can't be done ie. check that the aa step definitions are not global across all tests ?

File: Rakefile [hunk 6]
```diff
@@ -7,4 +7,4 @@ Osra::Application.load_tasks
 
 require 'coveralls/rake/task'
 Coveralls::RakeTask.new
-task :test_with_coveralls => [:spec, :cucumber, 'coveralls:push']
+task :test_with_coveralls => [:spec, :cucumber_hq, :cucumber_aa, 'coveralls:push']
```
Comment (NikitaAvvakumov) [inline Rakefile]: If people can give their feedback on #269, we can decide how best to handle this.

**Comment by NikitaAvvakumov** [comment]
For another PR: How about configuring test suites that would test either the entire application or only the the AA or HQ portions?
```bash
# full RSpec & Cucumber suites
rake test_all
# RSpec & Cucumber for AA only:
# - features/admin
# - all specs except spec/controllers/hq, spec/views, spec/routes/hq
rake test_aa
# RSpec & Cucumber for HQ only:
# - features/hq
# - all specs except spec/controllers/admin, spec/routes/admin
rake test_hq
```

File: features/hq/hq_features/authentication.feature [hunk 7]
```diff
@@ -0,0 +1,14 @@
+Feature:
+  As a site admin for osra
+  So that I have access to restricted pages
+  I would have to authenticate first
```
Comment (NikitaAvvakumov) [inline features/hq/hq_features/authentication.feature]: Since the key intent there is to show that protected pages _are not_ accessible without authentication, it may be more revealing to re-write this as something like
```
As an OSRA admin
So that I can be sure that the application is not accessible to the public
I would like to see that protected pages cannot be views without authentication
```

File: features/hq/hq_features/authentication.feature [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+  Scenario: Restricted pages should redirect unauthenticated users to login page
+    Given I am an unauthenticated user
+    When I go to the "partners" page for the "admin" role
```
Comment (NikitaAvvakumov) [inline features/hq/hq_features/authentication.feature]: Since we are testing authentication for the entire application, the partners page is not relevant per se. What do you think about abstracting it as something like `When I try to access a protected page`?

This can be even further abstracted as:
```cucumber
Given I am an unauthenticated visitor
Then I should not be able to access protected areas of the application
```

File: features/hq/hq_features/authentication.feature [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+    Then I should be on the "login" page for the "admin" role
+
+  Scenario: Restricted pages should be accessibe by authenticated users
+    Given I am a new, authenticated user
+    When I go to the "partners" page for the "admin" role
```
Comment (NikitaAvvakumov) [inline features/hq/hq_features/authentication.feature]: As above

File: features/hq/step_definitions/basic_steps.rb [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+Given(/^I am a new, authenticated user$/) do
+  create_admin_user
+  login
+end
+
+Given(/^I am an unauthenticated user$/) do
+  logout
+end
+
+When(/^I (?:go to|am on) the "([^"]*)" page for the "([^"]*)" role$/) do  |page, role|
```
Comment (NikitaAvvakumov) [inline features/hq/step_definitions/basic_steps.rb]: Now that we get to rebuild our entire Cucumber suite, let's have a discussion about whether it should use such low-level steps. I tend to think that Cucumber has more value when used at a [very high level of abstraction](http://lizkeogh.com/2011/03/04/step-away-from-the-tools/) (scroll to 'Write fewer scenarios').

File: features/hq/hq_features/authentication.feature (hunk 7 above)
Comment (razorcd) [inline features/hq/hq_features/authentication.feature]: Sounds much better. I am terrible at writing good feature sentences.

File: features/hq/hq_features/authentication.feature (hunk 8 above)
Comment (razorcd) [inline features/hq/hq_features/authentication.feature]: mmm... I agree it would sound better for the client. I'll change it.

File: features/hq/step_definitions/basic_steps.rb (hunk 10 above)
Comment (razorcd) [inline features/hq/step_definitions/basic_steps.rb]: That is interesting and it feels right.
1. We should only use BDD to test scenarios that bring value to the business of OSRA. (ex. authentication, create partners, sponsors, orpahns, start/terminate sponsorships, import orphan lists). This is somehow telling to not focus on "if every feature works" (happy/sad paths ...) and only focus on what the client needs this application for.

2. Instead of doing step by step features for every scenario we should use more general features and in step definitions use the helper methods (from step_helpers.rb) to keep the code DRY.

Example:
Instead of
```
Given I am a new, authenticated user
When I am on the "Partners" page
And I click "Add Partner"
And I fill in name with "John"
And I fill in city with "GB"
...
And I click "Create"
Then I should be on the "Partner" page for "John"
And I should see "Successfully created"
And I should see "John"
```

This scenario should look like:
```
Given I am a new, authenticated user
Then I should be able to create partners
```

or `I should be able to create partners from "create partners" page` if there are multiple places where partners can be created.

Then the step definition for `Then I should be able to create partners` should look like this:
```
Then(/^I should be able to create partners$/) do
  visit_partners_page
  click "Add Partner"
  expect(current_path).to be(new_partner_path)
  fill_in_partner_form_fields_with(:partner)
  click "Create partner"
  expect(current_path).to be(partner_path(1))
  expect(response).to have_content("Successfully created")
  expect(response).to have_content(partner.name)
end
```

PS: even if our client is not engaged in our cucumber tests and by that I mean we don't really need them, I think we should still build these tests (at a minimum/high level) just for practice/experience. And it also helps future new developers learn how OSRA works.

File: lib/tasks/cucumber.rake [hunk 11]
```diff
@@ -32,6 +32,18 @@ begin
       t.profile = 'rerun'
     end
 
+    Cucumber::Rake::Task.new({:aa => 'test:prepare'}, 'Run aa features that should pass') do |t|
```
Comment (razorcd) [inline lib/tasks/cucumber.rake]: It's the description of the task. if you run `$ rake -T` you can see all tasks with their descriptions.

File: config/cucumber.yml [hunk 12]
```diff
@@ -2,7 +2,13 @@
 rerun = File.file?('rerun.txt') ? IO.read('rerun.txt') : ""
 rerun_opts = rerun.to_s.strip.empty? ? "--format #{ENV['CUCUMBER_FORMAT'] || 'progress'} features" : "--format #{ENV['CUCUMBER_FORMAT'] || 'pretty'} #{rerun}"
 std_opts = "-r features --format #{ENV['CUCUMBER_FORMAT'] || 'pretty'} --strict --tags ~@wip"
+std_opts_aa = "-r features/aa -r features/support --format #{ENV['CUCUMBER_FORMAT'] || 'pretty'} --strict --tags ~@wip_aa"
+std_opts_hq = "-r features/hq -r features/support --format #{ENV['CUCUMBER_FORMAT'] || 'pretty'} --strict --tags ~@wip_hq"
 %>
```
Comment (razorcd) [inline config/cucumber.yml]: @PurityControl  I checked that only the specified spec definition files are loaded for aa and for hq.
Also, here is where we specify what folders should be loaded when `cucumber -p aa` or cucumber -p hq` is called.
Only the `support` folder is shared and that is loaded first by cucumber.

Comment (PurityControl) [inline config/cucumber.yml]: :+1:

**Comment by NikitaAvvakumov** [comment]
Looks ready for merge, but needs rebase

File: lib/tasks/cucumber.rake [hunk 13]
```diff
@@ -41,10 +53,12 @@ begin
       ::CodeStatistics::TEST_TYPES << "Cucumber features" if File.exist?('features')
     end
   end
-  desc 'Alias for cucumber:ok'
-  task :cucumber => 'cucumber:ok'
 
+  desc 'Alias for cucumber:ok'
   task :default => :cucumber
+  task :cucumber => 'cucumber:aa'
```
Comment (theB264) [inline lib/tasks/cucumber.rake]: I pulled your branch down and played with it.  Is there any way to change it so that if you do not specify `hq` or `aa` that it runs all the tests in the app instead of just `aa`?

Comment (razorcd) [inline lib/tasks/cucumber.rake]: you mean by running `cucumber` you want it to run the `hq` first and then the `aa` profiles?
That is only possible if we set a rake task for it.  like `$rake cucumber` that will invoke the 2 cucumber commands one after another. like: `$ cucumber -p hq && cucumber -p aa`

(The commands available at the moment are at the top of the page)

Comment (theB264) [inline lib/tasks/cucumber.rake]: Yes, precisely that.  Except I would run them in alphabetical order like cucumber does but that's extremely minor which order they run in.

**Comment by razorcd** [comment]
PR updated

I discarded the cucumber profile changes I made is earlier commits and configured everything only from rake tasks. It's cleaner this way.
I used `$rake test:hq` instead of `$rake test_hq` just because this is a convention by rake (using namespaces).

**Notice:**
- Model specs are run together with the aa spec tests!
- Cucumber tests are running in verbose mode: it lists all the loaded files when starts
- features/support is shared between the cucumber:aa and cucumber:hq tests
- features/support/step_helper.rb has the helper methods for hq step definitions 
(features/support folder is always loaded first by cucumber)



**Commands:**
- `$rake`    # -> will run all tests: spec(aa and hq), `cucumber:hq` and `cucumber:aa`
- `$rake test:aa` # -> will run all aa tests: `spec:aa`, `cucumber:aa`
- `$rake test:hq` # -> will run all hq tests: `spec:hq`, `cucumber:hq`
- `$rake spec:aa` # -> will run all aa specs: all specs that are not in a 'hq' folder
- `$rake spec:hq` # -> will run all hq specs: all specs that are in a 'hq' folder
- `$rake cucumber:aa` # -> will run all aa cucumber tests: all features from features/aa and will load all .rb files from  features/aa and features/support
- `$rake cucumber:hq` # -> will run all hq cucumber tests: all features from features/hq and will load all .rb files from  features/hq and features/support

PS: I'll remove the comments once this is ready to be merged.

File: Rakefile [hunk 14]
```diff
@@ -4,3 +4,44 @@
...
+    task :aa do
+      # sh 'rspec --exclude-pattern "spec/*/hq/**/*_spec.rb, spec/*/*/hq/**/*_spec.rb"'
+      # sh 'rspec spec/*/hq spec/*/*/hq'
+      sh 'rspec --exclude-pattern spec/\*\*\/hq{,/\*/\*\*\}/\*_spec.rb'  # model specs are also tested here here
+    end
+
+    task :hq do
+      # sh 'rspec --pattern "spec/*/hq/**/*_spec.rb, spec/*/*/hq/**/*_spec.rb"'
+      # sh 'rspec spec/*/hq spec/*/*/hq'
+      sh 'rspec --pattern spec/\*\*\/hq{,/\*/\*\*\}/\*_spec.rb'
```
Comment (NikitaAvvakumov) [inline Rakefile]: I think that model specs need to run here too. This way, the entirety of the hq application can be tested independently of the aa portion.

Comment (razorcd) [inline Rakefile]: Yes, I agree. don't know why I didn't add them from the beginning.
`$rake` is running all the steps (aa and hq) in one set, so the model specs will not run twice.

Comment (theB264) [inline Rakefile]: What do you think about running the specs with `rspec -fd -b`?  Or do you like the dots

Comment (razorcd) [inline Rakefile]: I do like the dots but lets discuss in scrum and vote.

**Comment by razorcd** [comment]
PR updated: `$rake spec:hq` also runs the model specs too
